### ISSUE TITLE
fibonacci - User might not be able to finalize their transaction

### ISSUE THREAD

**Issue by sherlock-admin2** [sherlock-audit/2024-02-optimism-2024-judging#119]
fibonacci

medium

# User might not be able to finalize their transaction

## Summary

There might be a case when a user needs to prove their transaction again but cannot do so, and as a result, cannot finalize it.

## Vulnerability Detail

A user can prove their transaction just after a dispute game is created, while it is still in the `IN_PROGRESS` status.

There might be cases when the user needs to prove their transaction again (for example, if the game is blacklisted or if the `respectedGameType` is changed).

Additionally, there might be different game types, and the `respectedGameType` can be changed and rolled back afterwards.

An example provided by a protocol team member:

> Let's say the fault proof system failed and we needed to fall back to the `PermissionedDisputeGame`, which follows the same security model as the OP Stack today with permissioned actors who are allowed to propose and challenge outputs.

> In this case, we'd change the respected game type to `2` from `1`. We have the ability to upgrade the implementation of game type `1` through the `DisputeGameFactory`'s `setImplementation` function as well. Once the bug in the permissionless game (`FaultDisputeGame.sol`) was resolved, we'd be able to change the respected game type in the `OptimismPortal` back to `1`. Because of the `respectedGameTypeUpdatedAt` timestamp check, it is intended that games that were created prior to rolling back would not be considered valid in the context of being able to finalize withdrawals that were proven against them.

Now let's examine the following scenario:

- A user proves a withdrawal transaction but does not finalize it yet.
- The `respectedGameType` is changed (e.g due to a bug in the `FaultDisputeGame` implementation).
- The `respectedGameType` is reverted back to its previous value (e.g. once the bug is resolved).
- During these updates, the game with which the user proved their transaction is resolved with the `DEFENDER_WINS` status.
- The user is now unable to finalize the transaction because the `respectedGameTypeUpdatedAt` timestamp was updated, and the creation time of the game with which the user proved their transaction is earlier.
- The user needs to re-prove their transaction to finalize it.
- However, the user cannot do this because none of the conditions for this are met.

```solidity
require(
    provenWithdrawal.timestamp == 0 || oldGame.status() == GameStatus.CHALLENGER_WINS
    || disputeGameBlacklist[oldGame] || oldGame.gameType().raw() != respectedGameType.raw(),
    "OptimismPortal: withdrawal hash has already been proven, and the old dispute game is not invalid"
);
```

- `provenWithdrawal.timestamp > 0` - the user has already proved the transaction before.
- `oldGame.status() == DEFENDER_WINS` - the game was resolved in favor of the defender.
- The dispute game was not blacklisted, and the `respectedGameType` is the same.

## Impact

This issue causes valid withdrawals to temporarily become non-finalizable.

## Code Snippet

https://github.com/sherlock-audit/2024-02-optimism-2024/blob/main/optimism/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L287-L291

## Tool used

Manual Review

## Recommendation

```diff
         require(
             provenWithdrawal.timestamp == 0 || oldGame.status() == GameStatus.CHALLENGER_WINS
-                || disputeGameBlacklist[oldGame] || oldGame.gameType().raw() != respectedGameType.raw(),
+                || disputeGameBlacklist[oldGame] || oldGame.gameType().raw() != respectedGameType.raw()
+                || oldGame.createdAt().raw() < respectedGameTypeUpdatedAt,
             "OptimismPortal: withdrawal hash has already been proven, and the old dispute game is not invalid"
         );
```

**Comment by 0xf1b0** [comment]
Escalate

> I don't see a reason why the guardian would revert the game type back.

I asked the sponsor in the private thread if the game type is supposed to change back to its previous value. Please check the Vulnerability Details section; I've included their response, which outlines a potential scenario.

It specifically states:

> we'd be able to change the respected game type in the `OptimismPortal` back

I don't understand why their opinion changed. However, according to the Sherlock [rules](https://docs.sherlock.xyz/audits/judging/judging#iii.-sherlocks-standards), if the contest's README or documentation do not explicitly state otherwise, the protocol's answers are considered the last source of truth in the hierarchy. Therefore, I believe that this case should be considered valid, and accordingly, the issue as well.

I also believe that issues #169 and #186 are not duplicates. They describe the scenario where proven but not finalized withdrawals need to be re-proven if the respected game type is changed. This is intended as a design choice.

Issue #71 is similar but slightly different. It describes a case where a guardian updates the respected game type to the same value as the current one. It appears that this should be classified as an admin mistake.

**Comment by sherlock-admin2** [comment]
> Escalate
> 
> > I don't see a reason why the guardian would revert the game type back.
> 
> I asked the sponsor in the private thread if the game type is supposed to change back to its previous value. Please check the Vulnerability Details section; I've included their response, which outlines a potential scenario.
> 
> It specifically states:
> 
> > we'd be able to change the respected game type in the `OptimismPortal` back
> 
> I don't understand why their opinion changed. However, according to the Sherlock [rules](https://docs.sherlock.xyz/audits/judging/judging#iii.-sherlocks-standards), if the contest's README or documentation do not explicitly state otherwise, the protocol's answers are considered the last source of truth in the hierarchy. Therefore, I believe that this case should be considered valid, and accordingly, the issue as well.
> 
> I also believe that issues #169 and #186 are not duplicates. They describe the scenario where proven but not finalized withdrawals need to be re-proven if the respected game type is changed. This is intended as a design choice.
> 
> Issue #71 is similar but slightly different. It describes a case where a guardian updates the respected game type to the same value as the current one. It appears that this should be classified as an admin mistake.

You've created a valid escalation!

To remove the escalation from consideration: Delete your comment.

You may delete or edit your escalation comment anytime before the 48-hour escalation window closes. After that, the escalation becomes final.

**Comment by alexfilippov314** [comment]
According to [Sherlock's duplication rules](https://docs.sherlock.xyz/audits/judging/judging#ix.-duplication-rules):
```
Scenario A:
There is a root cause/error/vulnerability A in the code. This vulnerability A -> leads to two attack paths:
- B -> high severity path
- C -> medium severity attack path/just identifying the vulnerability.
Both B & C would not have been possible if error A did not exist in the first place. In this case, both B & C should be put together as duplicates.
```

I agree that the scenario outlined in my issue is less likely, but the root vulnerability is clearly the same. That's why I believe the issue #71 is a valid duplicate.

**Comment by nevillehuang** [comment]
I believe this issue falls under the following sherlock category for trusted admin permissions. Additionally, there is potentially no loss of funds here given withdrawals can be reproven with a different address.

> An admin action can break certain assumptions about the functioning of the code. Example: Pausing a collateral causes some users to be unfairly liquidated or any other action causing loss of funds. This is not considered a valid issue.

**Comment by Evert0x** [comment]
I believe the language in the README is clear enough that only issues related to "The Guardian must not be able to:" are in-scope.

This was also confirmed by the protocol team on Discord https://discord.com/channels/812037309376495636/1222560798048518216/1224428452682399774

> If you can find a way to make the guardian perform any actions under the The Guardian must not be able to section, this is directly in scope

Planning to reject escalation and leave issue as is

**Comment by 0xf1b0** [comment]
Hi, @nevillehuang @Evert0x what about this part of the README?

> Additionally, there should be no function that can prevent valid withdrawals from being finalized (at most Medium).

The user is unable to finalize their transaction, and this situation arises not due to the actions of the guardian, but because of the state where current verifications does not allow a withdrawal to be re-proved.

> Withdrawals can be reproven with a different address.

Yes, but the user is unaware of this possibility, as they most likely do not review the code.

**Comment by Evert0x** [comment]
@0xf1b0 where is this sentence located in the readme?

> Additionally, there should be no function that can prevent valid withdrawals from being finalized (at most Medium).

**Comment by 0xf1b0** [comment]
> @0xf1b0 where is this sentence located in the readme?
> 
> > Additionally, there should be no function that can prevent valid withdrawals from being finalized (at most Medium).

<img width="843" alt="Screenshot 2024-05-02 at 21 04 38" src="https://github.com/sherlock-audit/2024-02-optimism-2024-judging/assets/145198103/647f845b-fd21-4d92-8b7d-87d16119933f">

**Comment by trust1995** [comment]
This one is clearly a design decision to err on the side of caution! 
They are simply not interested in ever finalizing a withdrawal that at some point in time it's game stopped being respected. For example maybe it is not watched and guarded properly, etc. The intention of this sentence is they are interested in griefing attacks that can screw another person's withdrawal.

Funds are never at risk, and withdrawals can always be reproven.

**Comment by 0xf1b0** [comment]
The design decision is to not allow the finalization of withdrawals when the game type is no longer respected. If the respected type changes, the user can re-prove the withdrawal.

```solidity
oldGame.gameType().raw() != respectedGameType.raw()
```

However, in the scenario described above, the user cannot finalize the withdrawal with the currently respected type because it has been changed and then rolled back.

> Funds are never at risk,

This does not invalidate the issue, and that's why it's considered medium. Otherwise, it should be classified as high.

> and withdrawals can always be reproven.

This issue demonstrates the exact opposite.

**Comment by trust1995** [comment]
Withdrawals can always be proven from another account, is the intention.

You are repeating the explanation, but my argument holds. Let's wait for verdict.

**Comment by alexfilippov314** [comment]
Just some thoughts:
* There might be some external integrations that can't simply use another account, as the proving transaction is part of some other business logic.
* How would users react if they found themselves unable to finalize and reprove their withdrawal using the usual flow? I don't think they will be happy with the explanation: just use another account, it's the intention.

I believe the correct intention should be: users are always able to either eventually finalize or reprove their withdrawal using the same address.

**Comment by Evert0x** [comment]
I don't believe this reports identifies an issue. It recommends a change to the contracts to allow for a different workflow. 

Will reject escalation and keep issue as is

**Comment by alexfilippov314** [comment]
Hi, @Evert0x!

Just to provide a bit more context: Optimism actually fixed this issue [here](https://github.com/ethereum-optimism/optimism/pull/10154). And they employ similar reasoning to what's described in this issue: "users should be able to re-prove at any point."

**Comment by alexfilippov314** [comment]
And I think it clearly demonstrates that this issue is not about allowing for a different workflow, but rather about supporting the intended workflow.

**Comment by 0xf1b0** [comment]
@Evert0x could you please clarify how the recommended mitigation changes the workflow?

**Comment by nevillehuang** [comment]
@Evert0x 

I agree with your statements [here](https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119#issuecomment-2085576454). The guardian can explicitly switch games to decide whether or not that specifc game type can be used for withdrawals (this includes both proving and finalizing withdrawals). It does not violate what the guardian role cannot do per my comments [here](https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119#issuecomment-2057618068). Additionally, withdrawals can always be re-proven via another address.

The watson(s) are also assuming the game type will be switched back by the trusted guardian via information shared privately as noted in this comment [here](https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119#issuecomment-2074124290), which doesn't apply since this information wasn't made known publicly that this was the intention of the guardian role.

**Comment by alexfilippov314** [comment]
Hi, @Evert0x!

The issue is not about invalidating withdrawal proofs. It's about a deadlock situation where users can't finalize AND reprove their withdrawal using the same address.

> Additionally, withdrawals can always be re-proven via another address.

As I've said earlier there might be some external integrations that can't simply use another account, as the proving is part of some other business logic. Assuming that reproving is always possible from the same address (even if the game type was updated) is okay, as it aligns with how the protocol is intended to work.

Regarding information from private thread. Could you please help me understand what Watson(s) should do in this situation?

1. Watson(s) identified a potential edge case where the protocol's core functionality breaks.
2. Watson(s) didn't find an explicit description of this edge case in the README/docs/etc.
3. Now, Watson(s) ask the sponsor about this edge case in the private thread. I want to highlight that this is the only option for Watson(s) since asking this question in a public space will lead to an increasing number of duplicates.
4. Watson(s) got issues rejected because his issue is based on the infromation from the private thread.

So my question is, what should Watson(s) do in this situation? I'm new to Sherlock, so I just want to understand what I should do to operate within the rules and at the same time do my best to secure the protocol.

As additional notes, I want to share two unhealthy incentives that the current system creates:
* Sponsors may exploit this system to minimize the number of listed issues in the final report. It is obvious that sponsors have an incentive to act this way.
* Watson(s) are disincentivized to submit issues at all in this case.

**Comment by alexfilippov314** [comment]
> It does not violate what the guardian role cannot do per my comments https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119#issuecomment-2057618068.

It does violate this invariant from this [handbook](https://oplabs.notion.site/Public-OP-Stack-Fault-Proofs-Sherlock-Competition-Handbook-e4cfdf210a5c45c79230af19653163cc#f4c77ba3c4e94447bf6bda45eefe24f4) (at least for temporary case) though:
> Users must be able to execute messages that have actually been sent and included in the state of the L2ToL1MessagePasser contract on the L2 blockchain. Users must not be permanently prevented from executing a valid message (temporary blocks are valid findings but less severe).

**Comment by Evert0x** [comment]
@alexfilippov314 please use discord to ask generic questions about the process.

> The watson(s) are also assuming the game type will be switched back by the trusted guardian via information shared privately as noted in this comment https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119#issuecomment-2074124290, which doesn't apply since this information wasn't made known publicly that this was the intention of the guardian role.

Based on this statement. I will keep the issue as is.

**Comment by alexfilippov314** [comment]
@Evert0x, thank you!

I finally found a statement in the public Optimism [specs ](https://specs.optimism.io/experimental/fault-proof/stage-one/bridge-integration.html)that proves that Optimism considered the case described in the issue as possible:
> In the event that we'd like to switch back to permissioned proposals, we can change the RESPECTED_GAME_TYPE in the OptimismPortal to a deployment of the PermissionedFaultDisputeGame.

I guess now this argument can't be considered as valid.

**Comment by nevillehuang** [comment]
@alexfilippov314 Thanks for searching the information, this could make the issue possibly valid.

However, I'm not sure if the information is indicative that the switch from non-permissioned -> permissioned -> non-permissioned game type must occur. Also there is conflicting statements from below. 

It seems to me that guardian is explicitly stated to be trusted to change any game type to block any gametype to be used for withdrawals. So if they choose to switch game type then it is expected that withdrawals for that game type cannot be finalized

> Trigger the OptimismPortal2.setRespectedGameType function to change which game type can be used for withdrawals.

Once the game type is initially switched (due to a bug etc..), then the current withdrawals that are not finalized would no longer be deemed as valid. 

> Additionally there should be no function that can cause valid withdrawals to not be finalizable (at most Medium) except for the case that the “resolve” function sets the result to CHALLENGER_WINS (invalid proposal) when it should be DEFENDER_WINS.

The [code comments](https://github.com/sherlock-audit/2024-02-optimism-2024/blob/main/optimism/packages/contracts-bedrock/src/L1/OptimismPortal2.sol#L282-L285) also doesn't mention that re-proving is allowed for such cases (possibility of switching of game type back to original game type)

**Comment by alexfilippov314** [comment]
Hi, @nevillehuang! Thank you!

I've tried to summarize all the arguments for the issue's validity.

**Likelihood:**
While it's true that the switch from non-permissioned to permissioned to non-permissioned game types might not occur, the purpose of the audit is to ensure that the system works properly in any possible scenario, especially in scenarios where the game implementation has bugs. Arguably, the first step Optimism might take upon finding a bug in the game implementation is to switch the game type from non-permissioned to permissioned, ensuring that only trusted actors participate in the game and thereby forbidding any malicious behavior. After fixing the bug, they might update the non-permissioned game implementation and switch back to non-permissioned game. This approach aligns with information from the protocol team (from the private thread in the issue description) and is also marked as a possible scenario in the public Optimism specs. So, we have many sources (arguably logic, private thread, public specs) indicating that this scenario is possible and only one source of the opposite (the first comment in this issue), which can't be considered a valid source according to Sherlock rules, if I understand correctly.

Therefore, I believe the likelihood is closer to Medium, considering that the main purpose of this contest is to ensure that everything works correctly even if the game implementation has bugs.

**Impact:**
There is no problem with invalidating withdrawal proofs by updating the game type; it is clearly intended behavior. The issue arises from disallowing reproving withdrawals using the same address in the case described in the issue, which is not a design decision but rather a mistake that breaks (at least partially) core protocol functionality. Here are the arguments supporting this statement:
1. Regarding code comments: The code allows reproving using the same address in cases when the game type was updated to a new value. If there is a case when this behavior differs, this is a good reason to specify this exception in the comment. The absence of information about this particular case indicates more that this particular case was overlooked than that this case is an exception.
2. There is an invariant in the [handbook](https://oplabs.notion.site/Public-OP-Stack-Fault-Proofs-Sherlock-Competition-Handbook-e4cfdf210a5c45c79230af19653163cc#f4c77ba3c4e94447bf6bda45eefe24f4): "Users must be able to execute messages that have actually been sent and included in the state of the L2ToL1MessagePasser contract on the L2 blockchain. Users must not be permanently prevented from executing a valid message (temporary blocks are valid findings but less severe)." This issue violates this invariant because:
* It is clearly not obvious for normal users that they should use another account to prove withdrawals for their original account. I can't say for sure but there is a possibility that UI won't support this. The number of users able to construct calldata for the proving step by themselves is not significant.
* In some cases (external integration), withdrawal can't be proved using a different account because external integration can involve additional business logic on the proving step. So, at least in this case, withdrawals are blocked until Optimism updates the game type or fixes this bug.
3. Optimism [fixed ](https://github.com/ethereum-optimism/optimism/pull/10154)this issue right after the contest ended with the statement in the PR: "users should be able to re-prove at any point." Although this cannot be used as a solid argument since this information appeared after the contest, it clearly indicates what the Optimism team thinks about the intended behavior.
4. There is also an invariant in the Optimism [specs](https://specs.optimism.io/experimental/fault-proof/stage-one/bridge-integration.html): "Re-proving is allowed at any time by the user. When a withdrawal is re-proven, its proof maturity delay is reset." This indicates that it is intended behavior. I believe this invariant was updated after the contest since it wasn't correct at the time of the contest, but it again supports the issue's point.
5. From the logic perspective there are no arguments for this behaviour, I can't think of any reason why users should NOT be able to reprove their withdrawals using the same address in this particular case.

In summary, the impact of this issue is temporary blocks of withdrawal execution (it is definitely true for external integrations (contracts) with additional logic on the proving step, and it is arguably true for users since the workaround of using a different account is not obvious and may not be supported by the UI). For this particular impact, the Optimism team clearly stated in the handbook that it is a valid finding but less severe.

That's why overall, I think this issue is a valid Medium with Medium likelihood and Medium impact.

**Comment by JeffCX** [comment]
Hello,

if this is valid,

my submission https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/4

is a duplicate of this issue as well. thanks.

**Comment by Evert0x** [comment]
My current opinion is the following, although I will revisit it.

---

This issue shows the following impact "Additionally there should be no function that can prevent valid withdrawals from being finalized (at most Medium)." and it's dependend on the game going from non-permissioned -> permissioned -> non-permissioned. 

The spec contains the following language

> In the event that we'd like to switch back to permissioned proposals, we can change the RESPECTED_GAME_TYPE in the OptimismPortal to a deployment of the PermissionedFaultDisputeGame

I believe that action should be able to get reversed, allowing the bug to get triggered. I believe Medium severity is appriopriate.

**Comment by trust1995** [comment]
I'm afraid we've been been misled by a cups and balls trick to think the issue answers to the impact condition:
> Additionally there should be no function that can prevent valid withdrawals from being finalized (at most Medium).

We never prevent valid withdrawals from being finalized. After game type moves back and forth, the withdrawal is invalid by design (the last update time check). The end impact is: the function prevents _reproving_ a game with the respected game type that changed a type before, with the same prover address. That impact was never mentioned in scope, and logically speaking is extremely weak (on the weak side of Low). There's not a single way this could potentially affect funds, and the Optimism portal is designed so that it doesn't matter one bit who the prover is.

Likelihood has to also be on low side, because configuration is strictly in the hands of the owner, and we know of only one particular combo which may be relevant. Of course, this use case is also completely theoretical as FDGs are designed to be continuously generated as we move forward in time and new withdrawals are needed, creating new root hashes. Why an old hash needs to be reproven when a new one would include all previous unspent withdrawals, is beyond me.

To sum up, from a technical perspective it's a clear Low x Low = Low for me.

**Comment by Haxatron** [comment]
Hi, I apologise if this is off-topic, but I am not sure how else to bring this up. if this gets upgraded to medium can you take a look at #65 again which reports an issue of a similar nature, but was rejected on the grounds of users can reprove their withdraw transaction. @Evert0x @nevillehuang 

The preconditions for that one is that create dispute game must be called in the same block as prove withdrawal and therefore it is likely to happen when an external integration (contracts) call them in one transaction.

cc: @zigtur

**Comment by alexfilippov314** [comment]
Hi, @trust1995!

1. It is clear that you can't finalize a withdrawal with an invalidated current proof without reproving. I don't see how this rephrasing alters anything. It remains possible to reprove withdrawals in `OptimismPortal2` using a different address, but I've already presented my arguments for why this might be impractical in general (especially for external integrations) or highly inconvenient, and likely nearly impossible for regular users.

2. All the sources thay I've found and presented (including public ones) indicate that the sequence `non-permissioned -> permissioned -> non-permissioned` is an expected response to discovered bugs in the game implementation. Thus, this "particular combo" isn't just one of many; instead, it's the most probable course of action in the event of a bug in the game implementation. I believe I shouldn't justify that the likelihood of a bug in the game implementation is quite high, given the specifics of the contest.

3. Regarding this statement, "Why an old hash needs to be reproven when a new one would include all previous unspent withdrawals, is beyond me." I'm not entirely sure what you mean here. I presume you're referring to the old MPT root hash, but the problem lies in the fact that you can't reprove (with your address) using any root hash due to the reproving condition blocking you regardless. You might observe that this condition does not incorporate the root hash:
```solidity
require(
    provenWithdrawal.timestamp == 0 || oldGame.status() == GameStatus.CHALLENGER_WINS
        || disputeGameBlacklist[oldGame] || oldGame.gameType().raw() != respectedGameType.raw(),
    "OptimismPortal: withdrawal hash has already been proven, and the old dispute game is not invalid"
);
```

In summary, I don't see how any of your arguments counter mine or offer a new perspective on this issue. Based on your last statement, I get the impression that you believe users can reprove their withdrawals using a new MPT root in this case, but it is not true.

**Comment by trust1995** [comment]
> 1. It is clear that you can't finalize a withdrawal with an invalidated current proof without reproving. I don't see how this rephrasing alters anything. It remains possible to reprove withdrawals in `OptimismPortal2` using a different address, but I've already presented my arguments for why this might be impractical in general (especially for external integrations) or highly inconvenient, and likely nearly impossible for regular users. 

It matters because as explained, the stated impact Optimism was looking for is very different from the achieved impact. But let's let the judges come to a conclusion.

> 2\. All the sources say I've found and presented (including public ones) indicate that the sequence `non-permissioned -> permissioned -> non-permissioned` is an expected response to discovered bugs in the game implementation. Thus, this "particular combo" isn't just one of many; instead, it's the most probable course of action in the event of a bug in the game implementation. I believe I shouldn't justify that the likelihood of a bug in the game implementation is quite high, given the specifics of the contest.

Good point! At best this would give Med likelihood for a bug being found and Opt using that sequence. 

EDIT: Invalidated argument below, left for transparency

> Regarding this statement, "Why an old hash needs to be reproven when a new one would include all previous unspent withdrawals, is beyond me." I'm not entirely sure what you mean here. I presume you're referring to the old MPT root hash, but the problem lies in the fact that you can't reprove (with your address) using any root hash due to the reproving condition blocking you regardless. You might observe that this condition does not incorporate the root hash:

What you are missing is that there is a continuous stream of FDGs being played, each one on a **new** root hash. Without such stream of games, the L2->L1 would be stale, and that's the entire point of FDGs. Using a new FDG would require no reproving at all, we simply prove a newer FDG:
```
function proveWithdrawalTransaction(
    Types.WithdrawalTransaction memory _tx,
    uint256 _disputeGameIndex,
    Types.OutputRootProof calldata _outputRootProof,
    bytes[] calldata _withdrawalProof
```
That would pass the `provenWithdrawal.timestamp == 0` check and easily pass the `prove()` function, if valid.

There is no sensible reason to reprove a previous hash, and the contest scope does not mention that as an impact. Therefore the issue is purely theoretical (Informational).

**Comment by alexfilippov314** [comment]
@trust1995, I believe it's not me who is missing the point here. The `provenWithdrawal` is stored in the `provenWithdrawals` mapping by `withdrawalHash`:
```solidity
bytes32 withdrawalHash = Hashing.hashWithdrawal(_tx);
ProvenWithdrawal memory provenWithdrawal = provenWithdrawals[withdrawalHash][msg.sender];
```

And the `withdrawalHash` only uses information about the withdrawal and therefore doesn't involve the root hash:

```solidity
/// @notice Derives the withdrawal hash according to the encoding in the L2 Withdrawer contract
/// @param _tx Withdrawal transaction to hash.
/// @return Hashed withdrawal transaction.
function hashWithdrawal(Types.WithdrawalTransaction memory _tx) internal pure returns (bytes32) {
    return keccak256(abi.encode(_tx.nonce, _tx.sender, _tx.target, _tx.value, _tx.gasLimit, _tx.data));
}
```
So it doesn't matter which MPT root you try to use; the proving will still be blocked.

**Comment by trust1995** [comment]
@alexfilippov314 I stand corrected. Therefore I ask judges to only consider the other points raised.

**Comment by 0xlastByte** [comment]
@trust1995 can you look to this issue i don't know where is the problem with it https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/241

**Comment by Evert0x** [comment]
As it's not strictly violating the statement "Additionally there should be no function that can prevent valid withdrawals from being finalized (at most Medium)."

As valid withdrawals can always be finalized, in the worst case using another address. I believe it should be invalid.

This impact is both
a) UX inefficiency, the user is forced to switch between addresses to reprove and finalize their withdrawal
b) A potential issue in a future contract, as this currently non-existent future contract might not expect this behavior

For these impacts, the following rules apply

> User experience and design improvement issues: Issues that cause minor inconvenience to users where there is no material loss of funds are not considered valid. Funds are temporarily stuck and can be recovered by the administrator or owner. Also, if a submission is a design opinion/suggestion without any clear indications of loss of funds is not a valid issue.

> Future issues: Issues that result out of a future integration/implementation that was not mentioned in the docs/README or because of a future change in the code (as a fix to another issue) are not valid issues.

**Comment by alexfilippov314** [comment]
Just for transparency: A "minor" inconvenience is the requirement for the user to fully understand the protocol and be able to construct proving transactions themselves, including fetching and building the correct Merkle Patricia Tree proof, or to be able to correctly modify the Optimism Bridge UI, as the UI almost definitely will not support proving from another address because it just doesn't make a lot of sense.

**Comment by trust1995** [comment]
> Just for transparency: A "minor" inconvenience is the requirement for the user to fully understand the protocol and be able to construct proving transactions themselves, including fetching and building the correct Merkle Patricia Tree proof, or to be able to correctly modify the Optimism Bridge UI, as the UI almost definitely will not support proving from another address because it just doesn't make a lot of sense.

I've used the Blast bridge to withdraw from L2->L1 which as you know uses the Optimism setup, and easily through the UI used a different address to perform the proving (for OpSec reasons, I'm paranoid of sending data blobs from a high value wallet). Can send you the TXs in private.

That's just an anecdote to disprove your thesis, but regardless such UX concerns would fall out of scope as what matters is only on-chain security.

**Comment by 0xlastByte** [comment]
@Evert0x  what about this issue https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/241 this could prevent user from proving a valid withdrawal check this issue.

**Comment by alexfilippov314** [comment]
Well, it seems that I'm cornered in this case. I still believe that the requirement to use a different account (with available native currency to pay for gas) to prove withdrawal for your original account is not a "minor" inconvenience. But if it's only me who thinks this way, so be it. To me the only thing that matters is the security in general including availability as part of it.

**Comment by Evert0x** [comment]
Result:
Invalid
Has Duplicates

**Comment by sherlock-admin4** [comment]
Escalations have been resolved successfully!

Escalation status:
- [0xf1b0](https://github.com/sherlock-audit/2024-02-optimism-2024-judging/issues/119/#issuecomment-2074124290): rejected
